c++/src/ColumnWriter.cc

```diff
@@ -468,25 +472,1099 @@ namespace orc {
+
+    virtual void getColumnEncoding(
+        std::vector<proto::ColumnEncoding>& encodings) const override;
+
+    virtual void recordPosition() const override;
+
+  protected:
+    std::unique_ptr<RleEncoder> secRleEncoder, nanoRleEncoder;
+
+  private:
+    RleVersion rleVersion;
+    const Timezone& timezone;
+  };
+
+  TimestampColumnWriter::TimestampColumnWriter(
+                             const Type& type,
+                             const StreamsFactory& factory,
+                             const WriterOptions& options) :
+                                 ColumnWriter(type, factory, options),
+                                 rleVersion(RleVersion_1),
+                                 timezone(getLocalTimezone()){
+    std::unique_ptr<BufferedOutputStream> dataStream =
+        factory.createStream(proto::Stream_Kind_DATA);
+    std::unique_ptr<BufferedOutputStream> secondaryStream =
+        factory.createStream(proto::Stream_Kind_SECONDARY);
+    secRleEncoder = createRleEncoder(std::move(dataStream),
+                                     true,
+                                     rleVersion,
+                                     memPool);
+    nanoRleEncoder = createRleEncoder(std::move(secondaryStream),
+                                      false,
+                                      rleVersion,
+                                      memPool);
+
+    if (enableIndex) {
+      recordPosition();
+    }
+  }
+
+  static int64_t formatNano(int64_t nanos) {
+    if (nanos == 0) {
+      return 0;
+    } else if (nanos % 100 != 0) {
+      return (nanos) << 3;
+    } else {
+      nanos /= 100;
+      int64_t trailingZeros = 1;
+      while (nanos % 10 == 0 && trailingZeros < 7) {
+        nanos /= 10;
+        trailingZeros += 1;
+      }
+      return (nanos) << 3 | trailingZeros;
+    }
+  }
+
+  void TimestampColumnWriter::add(ColumnVectorBatch& rowBatch,
+                                  uint64_t offset,
+                                  uint64_t numValues) {
+    ColumnWriter::add(rowBatch, offset, numValues);
+
+    TimestampVectorBatch& tsBatch =
+      dynamic_cast<TimestampVectorBatch &>(rowBatch);
+
+    const char* notNull = tsBatch.hasNulls ?
+                          tsBatch.notNull.data() + offset : nullptr;
+    int64_t *secs = tsBatch.data.data() + offset;
+    int64_t *nanos = tsBatch.nanoseconds.data() + offset;
+
+    TimestampColumnStatisticsImpl* tsStats =
+        dynamic_cast<TimestampColumnStatisticsImpl*>(colIndexStatistics.get());
+    for (uint64_t i = 0; i < numValues; ++i) {
+      if (notNull == nullptr || notNull[i]) {
+        // TimestampVectorBatch already stores data in UTC
+        int64_t millsUTC = secs[i] * 1000 + nanos[i] / 1000000;
+        tsStats->increase(1);
+        tsStats->update(millsUTC);
+      } else if (!tsStats->hasNull()) {
+        tsStats->setHasNull(true);
+      }
+    }
+
+    for (uint64_t i = 0; i < numValues; ++i) {
```

Comment:
can this loop be fused with the above loop? Applicable to other Writers below as well?